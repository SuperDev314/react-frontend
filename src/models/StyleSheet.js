// @flow

import { cloneElement } from 'react'
import {
  IS_BROWSER,
  DISABLE_SPEEDY,
  SC_ATTR,
  SC_STREAM_ATTR,
} from '../constants'
import { makeTag, makeRehydrationTag, type Tag } from './StyleTags'
import extractComps from '../utils/extractCompsFromCSS'

/* determine the maximum number of components before tags are sharded */
let MAX_SIZE
if (IS_BROWSER) {
  /* in speedy mode we can keep a lot more rules in a sheet before a slowdown can be expected */
  MAX_SIZE = DISABLE_SPEEDY ? 40 : 1000
} else {
  /* for servers we do not need to shard at all */
  MAX_SIZE = -1
}

let sheetRunningId = 0
let global

class StyleSheet {
  id: number
  sealed: boolean
  forceServer: boolean
  target: ?HTMLElement
  tagMap: { [string]: Tag<any> }
  hashes: { [string]: string }
  deferred: { [string]: string[] }
  rehydratedNames: { [string]: boolean }
  tags: Tag<any>[]
  capacity: number
  clones: StyleSheet[]

  constructor(
    target: ?HTMLElement = IS_BROWSER ? document.head : null,
    forceServer?: boolean = false
  ) {
    this.id = sheetRunningId += 1
    this.sealed = false
    this.forceServer = forceServer
    this.target = forceServer ? null : target
    this.tagMap = {}
    this.hashes = {}
    this.deferred = {}
    this.rehydratedNames = {}
    this.tags = []
    this.capacity = 1
    this.clones = []
  }

  /* rehydrate all SSR'd style tags */
  rehydrate() {
    if (!IS_BROWSER || this.forceServer) {
      return this
    }

    const els = []
    const names = []
    let extracted = []
    let isStreamed = false

    /* retrieve all of our SSR style elements from the DOM */
    const nodes = document.querySelectorAll(`style[${SC_ATTR}]`)
    const nodesSize = nodes.length

    /* abort rehydration if no previous style tags were found */
    if (nodesSize === 0) {
      return this
    }

    for (let i = 0; i < nodesSize; i += 1) {
      // $FlowFixMe: We can trust that all elements in this query are style elements
      const el = (nodes[i]: HTMLStyleElement)

      /* check if style tag is a streamed tag */
      isStreamed = !!el.getAttribute(SC_STREAM_ATTR) || isStreamed

      /* retrieve all component names */
      const elNames = (el.getAttribute(SC_ATTR) || '').trim().split(/\s+/)
      const elNamesSize = elNames.length
      for (let j = 0; j < elNamesSize; j += 1) {
        const name = elNames[j]
        /* add rehydrated name to sheet to avoid readding styles */
        this.rehydratedNames[name] = true
        names.push(name)
      }

      /* extract all components and their CSS */
      extracted = extracted.concat(extractComps(el.textContent))
      /* store original HTMLStyleElement */
      els.push(el)
    }

    /* abort rehydration if nothing was extracted */
    const extractedSize = extracted.length
    if (extractedSize === 0) {
      return this
    }

    /* create a tag to be used for rehydration */
    const tag = makeTag(this.target, null, this.forceServer)
    const rehydrationTag = makeRehydrationTag(
      tag,
      els,
      extracted,
      names,
      isStreamed
    )

    /* reset capacity and adjust MAX_SIZE by the initial size of the rehydration */
    this.capacity = Math.max(1, MAX_SIZE - extractedSize)
    this.tags.push(rehydrationTag)

    /* retrieve all component ids */
    for (let j = 0; j < extractedSize; j += 1) {
      this.tagMap[extracted[j].componentId] = rehydrationTag
    }

    return this
  }

  /* retrieve a "global" instance of StyleSheet which is typically used when no other is available */
  static get global(): StyleSheet {
    return global || (global = new StyleSheet().rehydrate())
  }

  /* NOTE: This is just for backwards-compatibility with jest-styled-components */
  static get instance(): StyleSheet {
    return StyleSheet.global
  }

  /* reset the internal "global" instance */
  static reset(forceServer?: boolean = false) {
    global = new StyleSheet(undefined, forceServer).rehydrate()
  }

  /* adds "children" to the StyleSheet that inherit all of the parents' rules
   * while their own rules do not affect the parent */
  clone() {
    const sheet = new StyleSheet(this.target, this.forceServer)
    /* add to clone array */
    this.clones.push(sheet)

    /* clone all tags */
    sheet.tags = this.tags.map(tag => {
      const ids = tag.getIds()
      const newTag = tag.clone()

      /* reconstruct tagMap */
      for (let i = 0; i < ids.length; i += 1) {
        sheet.tagMap[ids[i]] = newTag
      }

      return newTag
    })

    /* clone other maps */
    sheet.rehydratedNames = { ...this.rehydratedNames }
    sheet.deferred = { ...this.deferred }
    sheet.hashes = { ...this.hashes }

    return sheet
  }

  /* force StyleSheet to create a new tag on the next injection */
  sealAllTags() {
    this.capacity = 1
    this.sealed = true
  }

  /* get a tag for a given componentId, assign the componentId to one, or shard */
  getTagForId(id: string): Tag<any> {
    /* simply return a tag, when the componentId was already assigned one */
    const prev = this.tagMap[id]
    if (prev !== undefined && !this.sealed) {
      return prev
    }

    let tag = this.tags[this.tags.length - 1]

    /* shard (create a new tag) if the tag is exhausted (See MAX_SIZE) */
    this.capacity -= 1
    if (this.capacity === 0) {
      this.capacity = MAX_SIZE
      this.sealed = false
      tag = makeTag(this.target, tag ? tag.styleTag : null, this.forceServer)
      this.tags.push(tag)
    }

    return (this.tagMap[id] = tag)
  }

  /* optimal caching: hash is known and name can be returned */
  getNameForHash(hash: string) {
    return this.hashes[hash]
  }

  /* rehydration check: name is known, hash is unknown, but styles are present */
  alreadyInjected(hash: any, name: string) {
    if (!this.rehydratedNames[name]) return false

    this.hashes[hash] = name
    return true
  }

  /* checks whether component is already registered */
  hasInjectedComponent(id: string): boolean {
    return !!this.tagMap[id]
  }

  /* registers a componentId and registers it on its tag */
  deferredInject(id: string, cssRules: string[]) {
    const { clones } = this
    for (let i = 0; i < clones.length; i += 1) {
      clones[i].deferredInject(id, cssRules)
    }

    this.getTagForId(id).insertMarker(id)
    this.deferred[id] = cssRules
  }

  inject(id: string, cssRules: string[], hash?: string, name?: string) {
    const { clones } = this
    for (let i = 0; i < clones.length; i += 1) {
      clones[i].inject(id, cssRules, hash, name)
    }

    /* add deferred rules for component */
    let injectRules = cssRules
    const deferredRules = this.deferred[id]
    if (deferredRules !== undefined) {
      injectRules = deferredRules.concat(injectRules)
      delete this.deferred[id]
    }

    const tag = this.getTagForId(id)
    tag.insertRules(id, injectRules)

    if (hash && name) {
      tag.names.push(name)
      this.hashes[hash] = name
    }
  }

  toHTML() {
    return this.tags.map(tag => tag.toHTML()).join('')
  }

  toReactElements() {
    const { id } = this

    return this.tags.map((tag, i) => {
      const key = `sc-${id}-${i}`
      return cloneElement(tag.toElement(), { key })
    })
  }
}

export default StyleSheet
