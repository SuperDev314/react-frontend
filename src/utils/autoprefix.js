import camelizeStyleName from 'fbjs/lib/camelizeStyleName'
import hyphenateStyleName from 'fbjs/lib/hyphenateStyleName'
import { autoprefix } from 'glamor/lib/autoprefix'

export default root => {
  root.walkDecls(decl => {
    const objStyle = { [camelizeStyleName(decl.prop)]: decl.value }
    const prefixed = autoprefix(objStyle)
    Object.keys(prefixed).reverse().forEach(newProp => {
      decl.cloneBefore({
        prop: hyphenateStyleName(newProp),
        value: prefixed[newProp],
      })
    })
    decl.remove()
  })
}
