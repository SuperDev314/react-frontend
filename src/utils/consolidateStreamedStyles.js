// @flow
/**
 * When using streaming rendering, style blocks are emitted in chunks directly.
 * This method used to be necessary to move all style elements before the
 * rehydration kicks in. However the rehydration now takes all SSR style tags,
 * combines them into a single tag, and adds it to the head.
 * This renders the consolidation useless
 */
export default function consolidateStreamedStyles() {
  /* noop */
}
