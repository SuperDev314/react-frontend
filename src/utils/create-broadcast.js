/**
 * Creates a broadcast that can be listened to, i.e. simple event emitter
 *
 * @see https://github.com/ReactTraining/react-broadcast
 */
const createBroadcast = (initialValue) => {
  let listeners = []
  let currentValue = initialValue

  return {
    publish(value) {
      currentValue = value
      listeners.forEach(listener => listener(currentValue))
    },
    subscribe(listener) {
      listeners.push(listener)

      // Publish to this subscriber once immediately.
      listener(currentValue)

      // eslint-disable-next-line no-return-assign
      return () =>
        listeners = listeners.filter(item => item !== listener)
    },
  }
}

export default createBroadcast
