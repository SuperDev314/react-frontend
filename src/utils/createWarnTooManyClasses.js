// @flow

import styleSheet from '../models/StyleSheet'

const LIMIT = 200

export default () => {
  let generatedClasses = {}
  let warningSeen = false

  return (className: string, displayName: string) => {
    if (!warningSeen) {
      generatedClasses[className] = true
      if (Object.keys(generatedClasses).length >= LIMIT) {
        const latestRule = styleSheet.rules().find(s => s.selectorText && s.selectorText.indexOf(`.${className}`) !== -1)
        // Unable to find latestRule in test environment.
        /* eslint-disable no-console, prefer-template */
        console.warn(`Over ${LIMIT} classes were generated for component ${displayName}. ` +
           'Consider using style property for frequently changed styles.\n' +
           'Example: const MyStyledComponent = styled(props => <div className={props.className} style={{color: props.color}}>{props.children}</div>)`width: 100px; height: 100px;`' +
           (latestRule ? `\nLatest generated class: ${latestRule.cssText}` : ''))
        warningSeen = true
        generatedClasses = {}
      }
    }
  }
}
