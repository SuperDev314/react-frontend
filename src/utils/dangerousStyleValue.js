// @flow
import unitless from '@emotion/unitless';

// Taken from https://github.com/facebook/react/blob/b87aabdfe1b7461e7331abb3601d9e6bb27544bc/packages/react-dom/src/shared/dangerousStyleValue.js
export default function dangerousStyleValue(name: string, value: any): any {
  // https://github.com/amilajack/eslint-plugin-flowtype-errors/issues/133
  // $FlowFixMe
  if (typeof value === 'symbol') {
    return value;
  }

  const isEmpty = value == null || typeof value === 'boolean' || value === '';
  if (isEmpty) {
    return '';
  }

  if (
    typeof value === 'number' &&
    value !== 0 &&
    !(unitless.hasOwnProperty(name) && unitless[name] === 1)
  ) {
    return `${value}px`; // Presumes implicit 'px' suffix for unitless numbers
  }

  return String(value).trim();
}
