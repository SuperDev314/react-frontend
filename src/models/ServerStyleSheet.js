// @flow
/* eslint-disable no-underscore-dangle */
import React from 'react'
import stream from 'stream'

import { IS_BROWSER, SC_STREAM_ATTR } from '../constants'
import StyleSheet from './StyleSheet'
import StyleSheetManager from './StyleSheetManager'

declare var __SERVER__: boolean

/* this error is used for makeStyleTag */
const sheetClosedErr =
  process.env.NODE_ENV !== 'production'
    ? `
Can't collect styles once you've consumed a ServerStyleSheet's styles!
ServerStyleSheet is a one off instance for each server-side render cycle.
- Are you trying to reuse it across renders?
- Are you accidentally calling collectStyles twice?
`.trim()
    : ''

const streamBrowserErr =
  process.env.NODE_ENV !== 'production'
    ? 'Streaming SSR is only supported in a Node.js environment; Please do not try to call this method in the browser.'
    : ''

export default class ServerStyleSheet {
  closed: boolean
  global: StyleSheet
  instance: StyleSheet

  constructor() {
    this.instance = StyleSheet.global.clone()
    this.closed = false
  }

  collectStyles(children: any) {
    if (this.closed) {
      throw new Error(sheetClosedErr)
    }

    return (
      <StyleSheetManager sheet={this.instance}>{children}</StyleSheetManager>
    )
  }

  getStyleTags(): string {
    if (!this.closed) {
      this.closed = true
    }

    return this.instance.toHTML()
  }

  getStyleElement() {
    if (!this.closed) {
      this.closed = true
    }

    return this.instance.toReactElements()
  }

  interleaveWithNodeStream(readableStream: stream.Readable) {
    if (!__SERVER__ || IS_BROWSER) {
      throw new Error(streamBrowserErr)
    }

    /* the tag index keeps track of which tags have already been emitted */
    const { instance } = this
    let instanceTagIndex = 0

    const streamAttr = `${SC_STREAM_ATTR}="true"`
    const ourStream = new stream.Readable()
    // $FlowFixMe
    ourStream._read = () => {}

    readableStream.on('data', chunk => {
      const { tags } = instance
      let html = ''

      /* retrieve html for each new style tag */
      for (; instanceTagIndex < tags.length; instanceTagIndex += 1) {
        const tag = tags[instanceTagIndex]
        html += tag.toHTML(streamAttr)
      }

      /* force our StyleSheets to emit entirely new tags */
      instance.sealAllTags()
      /* prepend style html to chunk */
      ourStream.push(html + chunk)
    })

    readableStream.on('end', () => {
      this.closed = true
      ourStream.push(null)
    })

    readableStream.on('error', err => {
      this.closed = true
      ourStream.emit('error', err)
    })

    return ourStream
  }
}
