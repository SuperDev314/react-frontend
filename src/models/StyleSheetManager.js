// @flow
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import StyleSheet, { CONTEXT_KEY } from './StyleSheet'
import ServerStyleSheet from './ServerStyleSheet'
import { tagConstructorWithTarget } from './BrowserStyleSheet'

class StyleSheetManager extends Component {
  sheetInstance: StyleSheet
  props: {
    sheet?: StyleSheet | null,
    target?: HTMLElement | null,
  }

  getChildContext() {
    return { [CONTEXT_KEY]: this.sheetInstance }
  }

  componentWillMount() {
    if (this.props.sheet) {
      this.sheetInstance = this.props.sheet
    } else if (this.props.target) {
      this.sheetInstance = new StyleSheet(
        tagConstructorWithTarget(this.props.target)
      )
    } else {
      throw new Error('StyleSheetManager expects either a sheet or target prop')
    }
  }

  render() {
    /* eslint-disable react/prop-types */
    // Flow v0.43.1 will report an error accessing the `children` property,
    // but v0.47.0 will not. It is necessary to use a type cast instead of
    // a "fixme" comment to satisfy both Flow versions.
    return React.Children.only((this.props: any).children)
  }
}

StyleSheetManager.childContextTypes = {
  [CONTEXT_KEY]: PropTypes.oneOfType([
    PropTypes.instanceOf(StyleSheet),
    PropTypes.instanceOf(ServerStyleSheet),
  ]).isRequired,
}

StyleSheetManager.propTypes = {
  sheet: PropTypes.oneOfType([
    PropTypes.instanceOf(StyleSheet),
    PropTypes.instanceOf(ServerStyleSheet),
  ]),
  target: PropTypes.shape({
    appendChild: PropTypes.func.isRequired,
  }),
}

export default StyleSheetManager
