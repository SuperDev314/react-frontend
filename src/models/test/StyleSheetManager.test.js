// @flow
/* eslint-disable react/no-multi-comp */
import React from 'react';
import { renderToString } from 'react-dom/server';
import { render } from 'react-dom';
import TestRenderer from 'react-test-renderer';
import StyleSheetManager from '../StyleSheetManager';
import ServerStyleSheet from '../ServerStyleSheet';
import StyleSheet from '../StyleSheet';
import { resetStyled } from '../../test/utils';
import Frame, { FrameContextConsumer } from 'react-frame-component';

let styled;
let consoleError;

const parallelWarning =
  'Warning: Detected multiple renderers concurrently rendering the same context provider. This is currently unsupported.';

describe('StyleSheetManager', () => {
  consoleError = console.error;

  beforeEach(() => {
    document.body.innerHTML = '';
    document.head.innerHTML = '';

    styled = resetStyled(true);

    jest
      .spyOn(console, 'error')
      .mockImplementation(msg => (msg !== parallelWarning ? consoleError(msg) : null));
  });

  it('should use given stylesheet instance', () => {
    const sheet = new ServerStyleSheet();
    const Title = styled.h1`
      color: palevioletred;
    `;
    renderToString(
      <StyleSheetManager sheet={sheet.instance}>
        <Title />
      </StyleSheetManager>
    );
    expect(sheet.getStyleTags().includes(`palevioletred`)).toEqual(true);
  });

  it('should render its child', () => {
    const target = document.head;

    const Title = styled.h1`
      color: palevioletred;
    `;
    const renderedComp = TestRenderer.create(
      <StyleSheetManager target={target}>
        <Title />
      </StyleSheetManager>
    );

    expect(() => renderedComp.root.findByType(Title)).not.toThrowError();
  });

  it('should append style to given target', () => {
    const target = document.body;
    const Title = styled.h1`
      color: palevioletred;
    `;
    class Child extends React.Component {
      render() {
        return <Title />;
      }
    }

    expect(document.body.querySelectorAll('style')).toHaveLength(0);

    TestRenderer.create(
      <StyleSheetManager target={target}>
        <Child />
      </StyleSheetManager>
    );

    const styles = target.querySelector('style').textContent;

    expect(styles.includes(`palevioletred`)).toEqual(true);
  });

  it('should append style to given target in iframe', () => {
    const iframe = document.createElement('iframe');
    const app = document.createElement('div');

    document.body.appendChild(iframe);
    iframe.contentDocument.body.appendChild(app);

    const target = iframe.contentDocument.head;
    const Title = styled.h1`
      color: palevioletred;
    `;

    class Child extends React.Component {
      render() {
        return <Title />;
      }
    }

    render(
      <StyleSheetManager target={target}>
        <Child />
      </StyleSheetManager>,
      app
    );

    const styles = target.querySelector('style').textContent;
    expect(styles.includes(`palevioletred`)).toEqual(true);
  });

  it('should apply styles to appropriate targets for nested StyleSheetManagers', () => {
    const ONE = styled.h1`
      color: red;
    `;
    const TWO = styled.h2`
      color: blue;
    `;
    const THREE = styled.h3`
      color: green;
    `;

    TestRenderer.create(
      <div>
        <ONE />
        <StyleSheetManager target={document.head}>
          <div>
            <TWO />
            <StyleSheetManager target={document.body}>
              <THREE />
            </StyleSheetManager>
          </div>
        </StyleSheetManager>
      </div>
    );

    expect(document.head.innerHTML).toMatchSnapshot();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  // https://github.com/styled-components/styled-components/issues/1634
  it('should inject styles into two parallel contexts', async () => {
    const Title = styled.h1`
      color: palevioletred;
    `;

    // Injects the stylesheet into the document available via context
    const SheetInjector = ({ children, target }) => (
      <StyleSheetManager target={target}>{children}</StyleSheetManager>
    );

    class Child extends React.Component {
      componentDidMount() {
        const styles = this.props.document.querySelector('style').textContent;
        expect(styles.includes(`palevioletred`)).toEqual(true);
        this.props.resolve();
      }
      render() {
        return <Title />;
      }
    }

    const div = document.body.appendChild(document.createElement('div'));

    let promiseA, promiseB;
    promiseA = new Promise((resolveA, reject) => {
      promiseB = new Promise((resolveB, reject) => {
        try {
          // Render two iframes. each iframe should have the styles for the child injected into their head
          render(
            <div>
              <Frame>
                <FrameContextConsumer>
                  {({ document }) => (
                    <SheetInjector target={document.head}>
                      <Child document={document} resolve={resolveA} />
                    </SheetInjector>
                  )}
                </FrameContextConsumer>
              </Frame>
              <Frame>
                <FrameContextConsumer>
                  {({ document }) => (
                    <SheetInjector target={document.head}>
                      <Child document={document} resolve={resolveB} />
                    </SheetInjector>
                  )}
                </FrameContextConsumer>
              </Frame>
            </div>,
            div
          );
        } catch (e) {
          reject(e);
          div.parentElement.removeChild(div);
        }
      });
    });
    await Promise.all([promiseA, promiseB]);
    div.parentElement.removeChild(div);
  });

  describe('ssr', () => {
    it('should extract CSS outside the nested StyleSheetManager', () => {
      const sheet = new ServerStyleSheet();
      const ONE = styled.h1`
        color: red;
      `;
      const TWO = styled.h2`
        color: blue;
      `;
      class Wrapper extends React.Component {
        state = {
          targetRef: null,
        };
        render() {
          return (
            <div
              ref={el => {
                this.setState({ targetRef: el });
              }}
            >
              {this.state.targetRef && (
                <StyleSheetManager target={this.state.targetRef}>
                  <TWO />
                </StyleSheetManager>
              )}
            </div>
          );
        }
      }

      const html = renderToString(
        <StyleSheetManager sheet={sheet.instance}>
          <div>
            <ONE />
            <Wrapper />
          </div>
        </StyleSheetManager>
      );
      const css = sheet.getStyleTags();

      expect(html).toMatchSnapshot();
      expect(css).toMatchSnapshot();
    });
  });

  it('should render styles in correct order when styled(StyledComponent) and StyleSheetManager is used', () => {
    const Red = styled.div`
      color: red;
    `;
    const RedChangedToBlue = styled(Red)`
      color: blue;
    `;
    const sheet = new StyleSheet();
    const App = () => (
      <StyleSheetManager sheet={sheet}>
        <RedChangedToBlue>I should be blue</RedChangedToBlue>
      </StyleSheetManager>
    );
    const attachPoint = document.body.appendChild(document.createElement('div'));
    render(<App />, attachPoint);
    // window.getComputedStyles would be perfect, but it seems that JSDOM
    // implementation of that function isn't complete, so need to work around
    // it.
    const source = document.documentElement.outerHTML;
    // regex in case test is run against minified CSS in the future
    const indexOfRedStyle = source.search('color:red');
    const indexOfBlueStyle = source.search('color:blue');
    expect(indexOfRedStyle).toBeGreaterThanOrEqual(0);
    expect(indexOfBlueStyle).toBeGreaterThanOrEqual(0);
    expect(indexOfBlueStyle).toBeGreaterThan(indexOfRedStyle);
  });
});
