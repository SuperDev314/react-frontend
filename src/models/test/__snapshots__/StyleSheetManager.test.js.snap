// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleSheetManager should apply styles to appropriate targets for nested StyleSheetManagers 1`] = `
"<style type=\\"text/css\\" data-styled-components=\\"d\\" data-styled-components-is-local=\\"true\\">
/* sc-component-id: sc-a */
.sc-a {}
.d{color:red;}</style><style type=\\"text/css\\" data-styled-components=\\"e\\" data-styled-components-is-local=\\"true\\">
/* sc-component-id: sc-b */

.e{color:blue;}</style>"
`;

exports[`StyleSheetManager should apply styles to appropriate targets for nested StyleSheetManagers 2`] = `
"<style type=\\"text/css\\" data-styled-components=\\"f\\" data-styled-components-is-local=\\"true\\">
/* sc-component-id: sc-c */

.f{color:green;}</style>"
`;

exports[`StyleSheetManager ssr should extract CSS outside the nested StyleSheetManager 1`] = `"<div data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"-1987894045\\"><h1 class=\\"sc-a c\\" data-reactid=\\"2\\"></h1><div data-reactid=\\"3\\"></div></div>"`;

exports[`StyleSheetManager ssr should extract CSS outside the nested StyleSheetManager 2`] = `
"<style type=\\"text/css\\" data-styled-components=\\"c\\" data-styled-components-is-local=\\"true\\">
/* sc-component-id: sc-a */
.sc-a {}
.c{color:red;}</style>"
`;
